apps/remote_control/lib/lexical/remote_control/api/proxy/records.ex

```diff
@@ -0,0 +1,19 @@
+defmodule Lexical.RemoteControl.Api.Proxy.Records do
+  import Record
+
+  defrecord :message, body: nil
+  defrecord :mfa, module: nil, function: nil, arguments: []
```

Comment:
After spending some time looking through `Proxy` and `Proxy.State`, I'm thinking that `mfa` is the wrong kind of record. `mfa` is this super generic thing that doesn't (in my opinion) enforce what we want, which is that the mfas should only ever contain certain calls that we're specifically equipped to handle. (The implementation of `collapse_commands` in the proxy state, for instance, just drops commands it doesn't expect.)

I think it would be more clear and actually simplify some code if we had records for each command we're expecting to proxy.

```elixir
defrecord :schedule_compile, [:project, :force?]
defrecord :compile_document, [:project, :document]
defrecord :reindex
defrecord :reindex_running?
defrecord :format, [:document]
```

Maybe the name of this module becomes `Command` instead of `Records` as well and then has a `run/1` that can run each kind of thing:

```elixir
def run(message(body: body)) do
  RemoteControl.Dispatch.broadcast(body)
end

def run(schedule_compile(project: project, force?: force?)) do
  RemoteControl.Build.schedule_compile(project, force?)
end

...
```

Then the `apply` call in `Proxy` is just replaced by `Command.run/1` and becomes:

```elixir
state
|> State.flush()
|> Enum.each(&Command.run/1)
```

Reply:
the mfa is a deeply internal API to proxy and Proxy.State, how will this abstraction leak out through the public APIs ? 

One of the issues you flagged earlier was that there were too many records. What is the advantage of making what are effectively mfa records with a 1:1 correspondence to functions elsewhere?

Reply:
so, i think that this suggestion will have a knock-on effect in `Proxy.State` that will make it more complicated. Rather than checking for `mfa()` records inside of flush() we'll now need to check for each of the command records. Correct?